### DISCUSSION TITLE
InvalidRequestError when versioned history example is incorporated into existing application

### DISCUSSION THREAD

**Discussion by kkirsche** [sqlalchemy/sqlalchemy#10744] [Usage Questions] [answered]
👋  I was wondering if anyone has run into the following error before:

```
sqlalchemy.exc.InvalidRequestError: When initializing mapper mapped class Request->requests, expression 'requests_security_exceptions' failed to locate a name ("name 'requests_security_exceptions' is not defined"). If this is a class name, consider adding this relationship() to the <class 'request_manager.models.requests.request.Request'> class after both dependent classes have been defined.
```

I'm running into this when I add https://docs.sqlalchemy.org/en/14/_modules/examples/versioned_history/history_meta.html to an existing application. Once I make a single class versioned, this occurs whether it's a new class or an existing class. 

I'm not sure where to start trying to troubleshoot this one, but I was wondering if anyone has tips on how to approach this. Without modifying the mapper as `_history_mapper` does, everything works wonderfully.

In this case, this is referencing a many-to-many relationship association table name which is provided in the `secondary` relationship attribute.

The relationship between the two classes is many-to-many with an association table. One Request can have multiple exceptions, and an exception may belong to multiple requests.

The classes are defined in roughly the following way:

**Request**
```python
# a ton of imports in this file

class Request(Base):
    __tablename__ = "requests"

    id: Mapped[int] = Column(
        Integer(), primary_key=True, comment="The primary key of a request."
    )
    created_at: Mapped[datetime] = Column(
        TZDateTime(timezone=True),
        nullable=False,
        default=TZDateTime.utcnow,
        server_default=func.current_timestamp(),
        comment="When the request was created.",
    )
    # more attributes...
    security_exceptions: Mapped[list[SecurityException]] = relationship(
        "SecurityException",
        uselist=True,
        secondary="requests_exceptions",
        cascade_backrefs=False,
    )
    # various utility methods come below, removed to save space / simplify
```

**Security Exception**
```python
from __future__ import annotations

from typing import Any

from sqlalchemy import Column, String
from sqlalchemy.orm import Mapped, validates

from request_manager.models.base import Base
from request_manager.utilities.type_adapters import SecurityExceptionTypeAdapter


class SecurityException(Base):
    __tablename__: str = "security_exceptions"

    id: Mapped[str] = Column(
        String(length=255),
        primary_key=True,
        nullable=False,
    )

    @validates("id")
    def validates_exception_id(
        self, key: str, exception_id: Any
    ) -> str:
        return SecurityExceptionTypeAdapter.validate_python(exception_id)
```

**Association Table**
```python
from __future__ import annotations

from typing import TYPE_CHECKING

from sqlalchemy import Column, ForeignKey, Integer
from sqlalchemy.orm import Mapped, relationship

from request_manager.models.base import Base
from request_manager.integrations.sqlalchemy import Versioned

if TYPE_CHECKING:
    from request_manager.models.requests import Request
    from request_manager.models.exceptions.security_exception import SecurityException


class RequestSecurityException(Base):
    __tablename__: str = "requests_security_exceptions"

    id: Mapped[int] = Column(
        Integer(),
        primary_key=True,
        autoincrement=True,
        nullable=False,
    )
    request_id: Mapped[int] = Column(
        ForeignKey("requests.id"),
        nullable=False,
    )
    security_exception_id: Mapped[int] = Column(
        ForeignKey("security_exceptions.id"),
        nullable=False,
    )

    request: Mapped[Request] = relationship(
        "Request",
        uselist=False,
        viewonly=True,
    )
    security_exception: Mapped[SecurityException] = relationship(
        "SecurityException",
        uselist=False,
        viewonly=True,
    )
```

Any tips for what direction to go with debugging this is appreciated. Thanks all, and if I figure it out I'll make sure to share an update

**Comment by zzzeek** [comment]
well I dont see any strings that are trying to interpret that name, which seems a little strange, that error would be if you were using the name "requests_security_exceptions" inside of an argument like `relationship(secondary="requests_security_exceptions")`.  the history meta might be causing intitialization to proceed sooner than is warranted perhaps.   

if you can take your mappings above, add in history meta and package it into either a single file, or a couple of files as it looks like module paths might be involved (try single file first) and get me a reproducer then we can see what the problem is.

**Comment by kkirsche** [comment]
Thanks, Mike. I'm working to put together an example that triggers it. I appreciate your help and time. 

What I found during debugging was that this error was being raised when the history mapper was running `iterate_properties` where I see it first when generating alembic migrations. `iterate_properties` calls `_check_configure`. This calls `_configure_registries` and eventually `_do_configure_registries` which has a comment:

```python
            # initialize properties on all mappers
            # note that _mapper_registry is unordered, which
            # may randomly conceal/reveal issues related to
            # the order of mapper compilation
```

I assume I have something wrong or the order of operations things are occurring is wrong, but I'll see if I can get an example for you either standalone or leveraging your existing [`test_versioning.py`](https://docs.sqlalchemy.org/en/14/_modules/examples/versioned_history/test_versioning.html)

I appreciate your time

**Reply by zzzeek** [reply]
you can trigger the configure step just using `Base.registry.configure()` directly, that's where the error is occurring

**Comment by kkirsche** [comment]
Ok, so I got a representative example where I'm seeing this. Below is the code that I've specifically added.

```python
# BEGIN: Kevin's code


Base = declarative_base()


class RequestsExceptions(Base, ComparableEntity):
    __tablename__ = "requests_exceptions"
    id = Column(Integer, primary_key=True)
    request_id = Column(ForeignKey("requests.id"))
    exception_id = Column(ForeignKey("exceptions.id"))


class SecurityException(Versioned, Base, ComparableEntity):
    __tablename__ = "exceptions"

    id = Column(Integer, primary_key=True)
    name = Column(String(50))
    requests = relationship(
        "Request", back_populates="exceptions", secondary="requests_exceptions"
    )


class Request(Versioned, Base, ComparableEntity):
    __tablename__ = "requests"

    id = Column(Integer, primary_key=True)
    name = Column(String(50))
    exceptions = relationship(
        "SecurityException",
        back_populates="requests",
        secondary="requests_exceptions",
    )


def main() -> None:
    engine = create_engine("sqlite:///:memory:", future=True)
    Base.metadata.create_all(engine)
    Session = sessionmaker(engine)
    with Session as sess:
        req = Request(name="Example Request")
        exc = SecurityException(name="Example Exception")
        req.exceptions.append(exc)
        sess.add(req)
        sess.add(exc)
        sess.commit()

        sess.delete(req)
        sess.commit()

        req2 = Request(name="sc2")
        sess.add(req2)
        sess.commit()

        RequestHistory = Request.__history_mapper__.class_

        # only one entry should exist in the history table; one()
        # ensures that
        reqdeleted = sess.query(RequestHistory).one()

        # If sc2 has the same id that deleted sc1 had,
        # it will fail when modified or deleted
        # because of the violation of the uniqueness of the primary key on
        # sometable_history
        ne_(req2.id, reqdeleted.id)


if __name__ == "__main__":
    main()
```

I've included the full code, stack trace, and requirements.txt file at:
https://gist.github.com/kkirsche/5bd88fbdbc1b6878e2045fdb0473df37

I've tried this with all classes `Versioned` and just the one I want versioned, but didn't find any differences.

EDIT: fix leftover typo

**Reply by kkirsche** [reply]
Fixed a typo, sorry about that

**Reply by zzzeek** [reply, marked as answer]
okey doke, your original examples were given in SQLA 2.0 format however this is the 1.4 version of the recipe...it looks like 2.0's version will have the same issue at the moment.

we have no tests for two classes related to each other that are both versioned, nobody has ever asked for this.

I'm not sure this works all the way, but I can get it to configure by just moving the _history_mapper() step to occur after the configure like this, however trying to patch this in 2.0, it's causing tests to fail

```py
class Versioned:
    use_mapper_versioning = True  # changed by kkirsche to True
    """if True, also assign the version column to be tracked by the mapper"""

    @declared_attr
    def __mapper_cls__(cls) -> Callable[..., Mapper]:
        def map_(cls, *arg, **kw) -> Mapper:
            mp = mapper(cls, *arg, **kw)

            @event.listens_for(mp, "mapper_configured")
            def _mapper_constructed(mapper, class_):
                _history_mapper(mapper)

            return mp

        return map_

    __table_args__ = {"sqlite_autoincrement": True}
    """Use sqlite_autoincrement, to ensure unique integer values
    are used for new rows even for rows that have been deleted."""
```

possibly the original example should use this form as well

**Reply by zzzeek** [reply]
OK that version wont work in 2.0 at least, for 1.4, this change will set "active_history" before the mappers are configured

```py
def _history_mapper(local_mapper: Mapper) -> None:
    cls = local_mapper.class_

    # set the "active_history" flag
    # on on column-mapped attributes so that the old version
    # of the info is always loaded (currently sets it on all attributes)
    for prop in local_mapper._props.values():
        attr = getattr(local_mapper.class_, prop.key)
        if hasattr(attr, "impl") and attr.impl is not None:
            attr.impl.active_history = True
        else:
            prop.active_history = True

```

we would need to support this case more formally in 2.0 / 2.1.  let me know if you are upgrading to 2.0 for this, as we can work on a more polished approach at that point.

**Reply by kkirsche** [reply]
I'll give that a shot, thank you!

RE: upgrading, we plan to upgrade this application to 2.0 as soon as we can, and have upgraded our other applications to 2.x. We currently have one dependency for this application that is preventing us from upgrading to 2.x so we're trying to address that. I can certainly help with this in the v2 branch once we're ready for the migration.

If it helps, the use case for what we're trying to do is essentially this, we would like to have a request that is versioned. The request has a variety of relationships, some one-to-one, some one-to-many, and a few many-to-many. Requests have various pieces of content in those other tables — security exceptions, approvals, instructions, etc. and we want to be able to ensure that at any version, we know the complete state related to that object at that time.

For example:

1. Req 1 created
3. Req 1 updated with a new status (v2 — status as in pending approval / in review)
4. Req 1 has an exception added to it
etc.

If it makes more sense to explicitly create a history object rather than relying on a dynamically created one to ensure all the relationships are mapped, I can certainly do that, but wanted to understand if that was the recommended approach or if relying on the ORM examples is best in this situation.

Thank you for your time and help, I really appreciate it.

**Reply by kkirsche** [reply]
Just working through getting this to generate the schemas via alembic but this worked perfectly in terms of getting things up and running. Thanks for your time and help 👍

**Reply by kkirsche** [reply]
Ah, ok — it's the event `mapper_configured` that was preventing alembic from seeing it 👍  

thanks again for your time!
